openupgrade_scripts/scripts/hr/16.0.1.1/post-migration.py

```diff
@@ -0,0 +1,116 @@
+        cr,
+        """
+        SELECT DISTINCT %(relation_comodel_field)s
+        FROM %(source_relation_table)s
+        WHERE %(relation_comodel_field)s IN (
+            SELECT %(relation_comodel_field)s
+            FROM %(source_relation_table)s
+            GROUP BY %(relation_comodel_field)s
+            HAVING COUNT(*) > 1
+        )
+        """
+        % {
+            "source_relation_table": source_relation_table,
+            "relation_comodel_field": relation_comodel_field,
+        },
+    )
+    for res in cr.fetchall():
+        _logger.error(
+            "hr.plan.activity.type(%s,) is linked to several hr.plan. "
+            "hr.plan.activity.type can only be linked to one hr.plan. "
+            "Fix these data before migrating to avoid data loss.",
+            res[0],
+        )
+
+
+def m2m_to_o2m(
+    env,
+    model,
+    field,
+    source_relation_table,
+    relation_source_field,
+    relation_comodel_field,
+):
+    """Transform many2many relations into one2many (with possible data
+    loss).
+
+    Use rename_tables in your pre-migrate script to keep the many2many
+    relation table and give them as 'source_relation_table' argument.
+
+    :param model: The target registery model
+    :param field: The field that changes from m2m to o2m
+    :param source_relation_table: The (renamed) many2many relation table
+    :param relation_source_field: The column name of the 'model' id
+    in the relation table
+    :param relation_comodel_field: The column name of the comodel id in
+    the relation table
+    """
+    columns = env[model]._fields.get(field)
+    target_table = env[columns.comodel_name]._table
+    target_field = columns.inverse_name
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE %(target_table)s AS target
+        SET %(target_field)s=source.%(relation_source_field)s
+        FROM %(source_relation_table)s AS source
+        WHERE source.%(relation_comodel_field)s=target.id
+        """
+        % {
+            "target_table": target_table,
+            "target_field": target_field,
+            "source_relation_table": source_relation_table,
+            "relation_source_field": relation_source_field,
+            "relation_comodel_field": relation_comodel_field,
+        },
+    )
+
+
+def drop_table(cr, table):
+    """Drop table
+
+    :param table: the name of the table that will be dropped
+    """
+    openupgrade.logged_query(
+        cr,
+        "DROP TABLE IF EXISTS %(table)s CASCADE" % {"table": table},
+    )
+
+
+@openupgrade.migrate()
+def migrate(env, version):
+    warn_about_dataloss(
+        env.cr,
+        "hr_plan_hr_plan_activity_type_rel",
+        "hr_plan_activity_type_id",
+    )
+    m2m_to_o2m(
+        env,
+        "hr.plan",
+        "plan_activity_type_ids",
+        "hr_plan_hr_plan_activity_type_rel",
+        "hr_plan_id",
+        "hr_plan_activity_type_id",
+    )
+    drop_table(env.cr, "hr_plan_hr_plan_activity_type_rel")
```

Comment:
Should we really drop the data?
It is always risky...

Reply:
I'm not sure either. @pedrobaeza or @legalsylvain do you have an opinion about the removal of a relation table for a Many2many field that is now a Many2one ?

Obviously, data will be lost in some cases. These cases are reported in the log by the `warn_about_dataloss()` function. But indeed, no way to retrieve the data after the migration. In my opinion, cleaning data should be done before migrating.

Reply:
One of the principles of OpenUpgrade is to preserve the previous information at maximum (unless it supposes a problem. Example: quant merging in v11 was a "destructive" operation, well more a transformative operation, but it was needed). In this case, if it's a problem to have the table with that name, rename it to `None` with the openupgradelib method (which put it an special name). If not, just leave it as is. `database_cleanup` will remove on your command later.

Reply:
You can also lift constraints with the 2 corresponding openupgradelib methods for avoiding relational problems. Check for examples in previous versions.

Reply:
@pedrobaeza Thank you. Then I will go that way. :)

Reply:
Done.